### PR TITLE
Academy

### PR DESCRIPTION
**Pull request by manangouhari**
### What does this PR do?
1. Implements the whole academy flow.
  - `/tezos/academy/`
  - `/tezos/academy/module-*`
  - `/tezos/academy/module-*/chapter-*`

2. Logic to sync progress both on backend and frontend(localStorage)

### How to test this manually?
1. Go to `/tezos/academy`
2. Make your way through the flow, do a few chapters.
3. Then, log in. This should batch update user's progress on the backend. Either check the console or network tab in devtools to see the api call and response to batch update progress.
4. Do a few more chapters. Progress should sync automatically with the backend when logged in.

### Additional context.
Using the PR URL for the backend right now, as the backend branch is not merged yet. Will update this later on.

### CODE REVIEW

**Comment by manangouhari** [comment]
@bhaskarSingh can you please check why the build is failing. 
It builds on my local.

**Comment by manangouhari** [comment]
@bhaskarSingh, I'm facing a weird bug with LiveEditor. It has become somewhat scrollable but it still gets weird when content overflows. Please take a look.

These are the files to look at - 
- src/templates/chapter/CodingInterfaceTaquito.js
- src/templates/chapterWithLiveEditor.js

**Comment by bhaskarSingh** [comment]
Try wrapping liveEditor with the given below div:

```
            <div
              style={{
                height: `${
                  buttonClicked
                    ? `calc(100vh - (${HeaderHeight} + ${FooterHeight} + ${ContractFileHeight} + ${OptionHeight} + ${OutputHeaderHeight} + ${OutputContentHeight}))`
                    : `calc(100vh - (${HeaderHeight} + ${FooterHeight} + ${ContractFileHeight} + ${OptionHeight}))`
                }`,
                width: `calc(100vw - (100vw / 2.4))`,
                overflowX: 'auto',
              }}
            >
              <LiveEditor
                style={{
                  minHeight: `${
                    buttonClicked
                      ? `calc(100vh - (${HeaderHeight} + ${FooterHeight} + ${ContractFileHeight} + ${OptionHeight} + ${OutputHeaderHeight} + ${OutputContentHeight}))`
                      : `calc(100vh - (${HeaderHeight} + ${FooterHeight} + ${ContractFileHeight} + ${OptionHeight}))`
                  }`,
                }}
              />
            </div>
```

**Comment by manangouhari** [comment]
> overflowX: 'auto',

Works like a charm.

**Comment by bhaskarSingh** [comment]
@manangouhari  Can you give a status update on this PR?

**Comment by bhaskarSingh** [comment]
@manangouhari  Can you resolve the build error?
Screenshot 2021-02-08 at 3 00 16 PM

Afterwards, I will resolve the merge conflicts with develop branch if any.

**Comment by bhaskarSingh** [comment]
@manangouhari build is failing again due to the below error.
Screenshot 2021-02-08 at 3 00 16 PM

**Comment by bhaskarSingh** [comment]
@manangouhari What all is left before this PR can be merged?

**Comment by manangouhari** [comment]
> @manangouhari What all is left before this PR can be merged?

1. Sync progress with backend on completing chapter.
2. Sync progress with backend on sign up
3. Implementing progress storage for Taquito chapter. This is essentially replicating what I did with `chapter-new.js` to `chapterWithLiveEditor.js`

I'm working on it rn.

**Comment by bhaskarSingh** [comment]
Great! I would probably need an overview over the call of what this PR does. There are 79 file changes which makes it hard to review it.

**Comment by manangouhari** [comment]
@bhaskarSingh please review and merge. I've updated the very first comment on this PR to explain what all is being accomplished in this PR.

File: src/components/SemiLiveProvider/index.js [hunk 1]
```diff
@@ -1,44 +1,39 @@
...
-    let _printlnBuffer = "";
-    function print(value) {
-      _printlnBuffer += value + "\\n";
-      render(_printlnBuffer);
-    }
+    // The following piece of code provides additional functionality
+    // to the user code such as print function and key import
+    const code = `
+    
     Tezos.setProvider({ rpc: 'https://api.tez.ie/rpc/carthagenet' });
```
Comment (bhaskarSingh) [inline src/components/SemiLiveProvider/index.js:37]: RPC needs to be updated to delphi testnet https://api.tez.ie/rpc/delphinet

Comment (manangouhari) [reply src/components/SemiLiveProvider/index.js:37]: Okay.

File: src/hooks/use-modules.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  `);
+
+  return data.allMdx.edges
+    .filter(module => module.node.frontmatter.slug == m)
+    .map(module => ({
+      module: module.node.frontmatter.slug,
+      title: module.node.frontmatter.title,
+    }))[0];
+
+  return ['working', 'on', 'it'];
```
Comment (manangouhari) [inline src/hooks/use-modules.js:27]: remove this.

File: src/templates/components/chapter/LearningInterface.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          <SubjectIcon />
+        </button>
+        <h5
+          className={`font-extrabold text-lg overflow-ellipsis overflow-hidden`}
+        >
+          {heading}
+        </h5>
+      </header>
+      <div className={`py-8 px-10 overflow-y-auto flex-1`}>
+        <MDXProvider
```
Comment (manangouhari) [inline src/templates/components/chapter/LearningInterface.js:22]: Fix rendering for code snippets.

**Comment by bhaskarSingh** [comment]
Todos left #162

**Pull request merged**